File: js/src/discover.js
import { isoToOrdinal, ordinalToIso } from './dates.js';

export const SEARCH_FLOOR = '2000-01-01';

/**
 * Yields 'YYYY-MM-DD' dates to probe, receives each date's availability via
 * next(bool), returns the earliest available date (null if nothing is
 * available). Assumes availability is a contiguous suffix: in-fault before
 * the floor, available from it onward.
 * @param {string} known known floor, 'YYYY-MM-DD'
 * @param {string} today 'YYYY-MM-DD'
 * @returns {Generator<string, string | null, boolean>}
 */
export function* earliestSearch(known, today) {
  const knownOrd = isoToOrdinal(known);
  let lo, hi;
  if (yield known) {
    if (!(yield ordinalToIso(knownOrd - 1))) return known;
    lo = isoToOrdinal(SEARCH_FLOOR);
    hi = knownOrd - 1;
  } else {
    hi = isoToOrdinal(today);
    if (!(yield today)) {
      hi -= 1;
      if (!(yield ordinalToIso(hi))) return null;
    }
    lo = knownOrd + 1;
  }
  while (lo < hi) {
    const mid = Math.floor((lo + hi) / 2);
    if (yield ordinalToIso(mid)) {
      hi = mid;
    } else {
      lo = mid + 1;
    }
  }
  return ordinalToIso(hi);
}

File: python/src/octaplus/_discover.py
from __future__ import annotations

from collections.abc import Generator
from datetime import date, timedelta

SEARCH_FLOOR = date(2000, 1, 1)

_ONE_DAY = timedelta(days=1)


def earliest_search(known: date, today: date) -> Generator[date, bool, date | None]:
    """Yields dates to probe, receives their availability, returns the earliest
    available date (None if nothing is available). Assumes availability is a
    contiguous suffix: in-fault before the floor, available from it onward."""
    if (yield known):
        if not (yield known - _ONE_DAY):
            return known
        lo, hi = SEARCH_FLOOR, known - _ONE_DAY
    else:
        hi = today
        if not (yield hi):
            hi = hi - _ONE_DAY
            if not (yield hi):
                return None
        lo = known + _ONE_DAY
    lo_o, hi_o = lo.toordinal(), hi.toordinal()
    while lo_o < hi_o:
        mid = (lo_o + hi_o) // 2
        if (yield date.fromordinal(mid)):
            hi_o = mid
        else:
            lo_o = mid + 1
    return date.fromordinal(hi_o)
